src/script/redmine-upkeep.py

```diff
@@ -0,0 +1,714 @@
+            return None
+
+    def add_or_update_custom_field(self, field_id, value):
+        """Helper to add or update a custom field in the payload."""
+        custom_fields = self.update_payload.setdefault("custom_fields", [])
+        found = False
+        current_value = self.get_custom_field(field_id) # Get current value from issue or payload
+
+        if current_value == value:
+            # Value is already the same, no change needed
+            log.debug(f"Field {field_id} is already set to '{value}'. No update needed.")
+            return
+
+        for field in custom_fields:
+            if field.get('id') == field_id:
+                field['value'] = value
+                found = True
+                break
+        if not found:
+            custom_fields.append({'id': field_id, 'value': value})
+        self.has_changes = True # Mark that a change has been made
+
+    def change_field(self, field, value):
+        self.update_payload[field] = value
+        self.has_changes = True
+
+    def get_update_payload(self, suppress_mail=True): # Added suppress_mail parameter
+        today = datetime.now(timezone.utc).isoformat(timespec='seconds')
+        self.add_or_update_custom_field(REDMINE_CUSTOM_FIELD_ID_UPKEEP_TIMESTAMP, today)
+
+        current_tags_str = self.get_custom_field(REDMINE_CUSTOM_FIELD_ID_TAGS)
+        current_tags = []
+        if current_tags_str:
+            current_tags = [tag.strip() for tag in current_tags_str.split(',') if tag.strip()]
+            if "upkeep-failed" in current_tags:
+                log.info(f"[Issue #%d] 'upkeep-failed' tag found in '%s'. Removing for update.", self.issue.id, current_tags_str)
+                current_tags.remove("upkeep-failed")
+                self.add_or_update_custom_field(REDMINE_CUSTOM_FIELD_ID_TAGS, ", ".join(current_tags))
+
+        payload = {
+            'issue': self.update_payload,
+        }
+        if suppress_mail:
+            payload['suppress_mail'] = "1"
+        return payload
+
+    def fetch_pr(self):
+        pr_id = self.get_custom_field(REDMINE_CUSTOM_FIELD_ID_PULL_REQUEST_ID)
+        if not pr_id:
+            log.debug("[Issue #%d] No PR ID found in 'Pull Request ID'.", self.issue.id)
+
+            # If not found in custom field, try to find it in the issue description
+            if self.issue.tracker.id == REDMINE_TRACKER_ID_BACKPORT and self.issue.description:
+                match = re.search(r'^https://github.com/ceph/ceph/pull/(\d+)$', self.issue.description.strip())
+                if match:
+                    pr_id = match.group(1)
+
+        try:
+            pr_id = int(pr_id)
+        except ValueError:
+            log.warning("[Issue #%d] Invalid PR ID '%s'.", self.issue.id, pr_id)
+            return None
+
+        if pr_id in self._pr_cache:
+            return self._pr_cache[pr_id]
+
+        endpoint = f"https://api.github.com/repos/{GITHUB_ORG}/{GITHUB_REPO}/pulls/{pr_id}"
```

Comment:
I'd put these 2 literal strings as globals as well. For the f-string, you'll need to use a Template one instead.